### ISSUE TITLE
RNG behaves differently in nested calls to Rcpp functions

### ISSUE THREAD

**Issue by wch** [RcppCore/Rcpp#823]
The global [`RNGScopeCounter`](https://github.com/RcppCore/Rcpp/blob/0069f0b0/src/api.cpp#L68), causes `GetRNGstate()` and `PutRNGstate()` to be called only when entering/exiting the lowest Rcpp function on the call stack. 

This can cause problems when  one Rcpp function calls another one, and the latter one uses the random number generator.

In the example below, the behavior of `f` is different when called directly from the console vs. called from another Rcpp function.

```R
library(Rcpp)

cppFunction('
  double rng() {
    return R::runif(1,10);
  }
')

cppFunction('
  Rcpp::RObject invoke(Rcpp::Function f) {
    return f();
  }
')

f <- function() {
  set.seed(100)
  a <- runif(1)
  rng()
  
  set.seed(100)
  rng()
  b <- runif(1)

  identical(a, b)
}

f()  # FALSE

invoke(f) # TRUE
```

cc: @jcheng5
From rstudio/shiny#1953.

**Comment by eddelbuettel** [comment]
As I discussed over there, I consider the example invalid and construed because of `Rcpp::Function()`.

**Comment by eddelbuettel** [comment]
I guess our basic guarantee is this one:

```r
R> set.seed(100)
R> c(rng(), runif(1))
[1] 0.307766 0.257673
R> set.seed(100)
R> c(runif(1), rng())
[1] 0.307766 0.257673
R> 
```

Using your `rng()` function (corrected using `R::runif(0,1)` to correspond to the R call -- ), we see that once a set is seed, it does matter whether your query the RNG via R or C.  That is a nice property.

It seems like piercing this guarantee with Rcpp::Function() to call back to R destroy it.  Oh well, so let's not do that then in apps relying on identical RNG streams.   One could always call out to other RNGs.

Slightly mod'ed example below:

```c++
#include <Rcpp.h>

// [[Rcpp::export]]
double rng() {
  return R::runif(0, 1);
}

// [[Rcpp::export]]
Rcpp::RObject invoke(Rcpp::Function f) {
  Rcpp::RObject z = f();
  return z;
}

/*** R
f <- function() {
  set.seed(100)
  a1 <- runif(1)
  b1 <- rng()

  set.seed(100)
  a2 <- rng()
  b2 <- runif(1)

  identical(c(a1, b1), c(a2, b2))
}

f()  

invoke(f) # not TRUE
*/
```

**Comment by kevinushey** [comment]
Some notes from a conversation with @wch (mostly writing here for my own benefit; this might already be obvious to those familiar with R's RNG).

`GetRNGstate()` and `PutRNGstate()` are effectively used to synchronize the R `.Random.seed` variable with the internal RNG state used by the active RNG. In effect:

- `GetRNGstate()` synchronizes `.Random.seed` => internal state;
- `PutRNGstate()` synchronizes internal state => `.Random.seed`.

In @wch's example, the call to `rng()` does not call `GetRNGstate()` / `PutRNGstate()`, so the C-level RNG table gets out of sync with `.Random.seed`. When a later usage of R's RNG is made (through e.g. `runif(1)`, the internal seed state is synchronized with the (now stale!) `.Random.seed`, and so it gets the the 'previous' result.

I think it should be safe to recursively call `GetRNGstate()` / `PutRNGstate()`, though. The only downside is we might end up with a couple unneeded calls to `GetRNGstate()` and `PutRNGstate()`. E.g. suppose we had ('unwrapping' the Rcpp function wrapper to show the calls to `GetRNGstate()` / `PutRNGstate()`):

```cpp
void f() {
    GetRNGstate();
        g();
        h();
    PutRNGstate();
}
```

In the above example, if `g()` makes use of the random number generator without calling `PutRNGstate()`, then `h()` (if it were to call `GetRNGstate()`) would end up with a stale view of the RNG. On the other hand:

```cpp
void f() {
    GetRNGstate();
	GetRNGstate();
	g();
	PutRNGstate();
	GetRNGstate();
	h();
	PutRNGstate();
    PutRNGstate();
}
```

The only real downside is that we're making a couple unnecessary calls to `GetRNGstate()` / `PutRNGstate()`. That is, repeated calls to `GetRNGstate()` aren't wrong, just wasteful -- whereas missing calls to `GetRNGstate()` / `PutRNGstate()` can definitely get the user into a bind.

tl;dr: I think it's safe and correct to call `GetRNGstate()` and `PutRNGstate()` recursively.

**Comment by eddelbuettel** [comment]
I'm mostly on the same page following the discussing with @wch and @jcheng5 over on the shiny ticket.
When I look at `git blame` though, I see that the counter-based call of the get/put functions was added by nobody other than ... you.  Do you recall details?   Just a simple optimization?

**Comment by kevinushey** [comment]
😬 

I cannot recall the details. Perhaps a misguided optimization on my part?

**Comment by coatless** [comment]
FYI this issue is similar to #268.

**Comment by eddelbuettel** [comment]
Good catch.  I still think the whole is somewhat secondary, but @kevinushey is of course influenced by what @wch and @jcheng5 have seen.  But to me the whole issue is still on the edge of overblown: it took years of usage an arguably large number of people ... to find a single unhappy shiny user who had somewhat non-standard code calling back to R.

As @jjallaire said in #268: _Calling R functions from within C++ code is not the design center_.  Couldn't agree more.  With the PR we will be making more RNG state change calls, all of which may tickle a garbage collection event as the discussion in #268 reminded us.  I still think that overall this is going to be bening ... but we don't really know.
